cuqi/solver/_solver.py

```diff
@@ -650,8 +650,148 @@ def solve(self):
                 x_new = x_new + ((k-1)/(k+2))*(x_new - x_old)
               
             x = x_new.copy()
+
+class ADMM(object):
+    """Alternating Direction Method of Multipliers for solving regularized linear least squares problems of the form:
+    Minimize ||Ax-b||^2 + sum_i f_i(L_i x).
+
+    Reference:
+    [1] Boyd et al. "Distributed optimization and statistical learning via the alternating direction method of multipliers."Foundations and Trends® in Machine learning, 2011.
+
+    
+    Parameters
+    ----------
+    A : ndarray or callable f(x,*args).
+    b : ndarray.
+    penalties : List of tuples (callable proximal operator of f_i, linear operator L_i).
+    x0 : ndarray. Initial guess.
+    tradeoff : Trade-off between linear least squares and regularization term in the solver iterates.
+    maxit : The maximum number of iterations.
+    adapative : Whether to adaptively update the tradeoff parameter each iteration. Based on [1], Subsection 3.4.1
     
+    Example
+    -----------
+    .. code-block:: python
     
+        from cuqi.solver import ADMM, ProximalL1, ProjectNonnegative
+        import scipy as sp
+        import numpy as np
+
+        rng = np.random.default_rng()
+
+        m, n = 10, 5
+        A = rng.standard_normal((m, n))
+        b = rng.standard_normal(m)
+        x0 = np.zeros(n)
+        admm = ADMM(A, b, x0, penalties = [(ProximalL1, np.eye(n)), (lambda z, _ : ProjectNonnegative(z))], tradeoff = 10)
+        sol, _ = admm.solve()
+
+    """  
+
+    def __init__(self, A, b, penalties, x0, tradeoff = 10, maxit = 100, inner_max_it = 10, adaptive = True):
+
+        self.A = A
+        self.b = b
+        self.x_cur = x0
+
+        dual_len = [penalty[1].shape[0] for penalty in penalties]
+        self.y_cur = [np.zeros(l) for l in dual_len]
+        self.u_cur = [np.zeros(l) for l in dual_len]
+        self.n = penalties[0][1].shape[1]
+        
+        self.rho = tradeoff
+        self.maxit = maxit
+        self.inner_max_it = inner_max_it
+        self.adaptive = adaptive
+
+        self.penalties = penalties
+       
+        self.p = len(self.penalties)
+        self._big_matrix = None
+
+    def solve(self):
+        y_new = self.p*[0]
+        u_new = self.p*[0]
+
+        # Iterating
+        for i in range(self.maxit):
+            big_matrix, big_vector = self._iteration_pre_processing()
```

Comment:
It seems that the new method here is storing self._big_matrix as an attribute in the class instance, but also working with a local variable "big_matrix" here. Perhaps it might be best to use just self._big_matrix to avoid potential confusion in which variable is being used. If needed you can also store self._big_vector and just use those. Then the method would be:
```python
self._iteration_pre_processing()
```